File: Assets/Scripts/TriggerHatch.cs

// Hannes Helmholz
//
// class for animated objects with one open and one close animation

using UnityEngine;

[RequireComponent (typeof(NetworkView))] // for RPC

public class TriggerHatch : MonoBehaviour
{

    // =============================================================================
    #region MEMBERS
    private Config Config;
    private NetworkView NetView;
	
	public TriggerButtonSequence Script;
	
	public enum TriggerType
    {
        Trigger,
        InstantOpen,
        PermanentOpen,
        PermanentClosed
    }
	public TriggerType Type;

    public enum TriggerObject
    {
        DistanceByCollider,
        Script
    }
	public TriggerObject Trigger;
 
	private bool IsInit;
   
    private Animation Ani;
	public int AnimationsOpenSize;
    public int AnimationsCloseSize;
    public AnimationClip[] AnimationsOpen;
    public AnimationClip[] AnimationsClose;
	
	
	private bool IsClose = true;
	

    
    #endregion
    // =============================================================================
 
//dan Konstructoren sind in unity ein noGo bei klassen die von MonoBehaviour erben !!!!
//http://answers.unity3d.com/questions/32413/using-constructors-in-unity-c.html	
//deshalb: constructor --> awake / start
//http://docs.unity3d.com/Documentation/ScriptReference/MonoBehaviour.Start.html
//http://docs.unity3d.com/Documentation/ScriptReference/MonoBehaviour.Awake.html
//awake wird immer asugeführt, start nur wen das gameobject aktiviert ist!
 
    // =============================================================================
    #region METHODS UNITY ---------------------------------------------------------------
 
	/**
	 * Diese Methode wird vor der Startmethode aufgerufen und setzt wichtige Parameter.
	 */
	void Awake()
	{		
		Type = TriggerType.PermanentOpen;
		Trigger = TriggerObject.Script;
		Config = GameObject.FindWithTag( "Config" ).GetComponent<Config>();
        Ani = animation;
        NetView = networkView;
	}
	/**
	 * Diese Methode wird einmal nach Awake aufgerufen.
	 */
	void Start()
	{
		if( !Config.IsServer && !Config.IsStandalone)
        {
            IsInit = true;
            return;
        }
     
        if( Type == TriggerType.Trigger )
        {
			IsInit = true;
        }
		
        if( Trigger == TriggerObject.Script )
        {
			Script = GetComponent<TriggerButtonSequence>();
        }
		//OnTriggerEnter(null);
		
	}
 
	/**
	 * Diese Methode wird jeden Update aufgerufen
	 */
    void Update()
    { 
        if( !Config.IsServer )
			return;
			
        if( !IsInit && Config.gameStarted )
        {
            MakePermanent();
        }
    }
 
 
	/**
	 * Diese Methode wird ausgelöst wenn ein Objekt den Trigger betritt.
	 */
    void OnTriggerEnter( Collider other )
    {
		if (IsClose)
		{
			// Handelt es sich dabei um den Spieler wird der Trigger ausgelöst.
			if (Config.IsStandalone && other.tag.Equals("Player")) {
				RPCAnimate(true);
			}
			else if (Config.IsServer){
				Open ();
			}
			IsClose = false;
		}
    }
 
 
	/**
	 * Diese Methode wird ausgelöst wenn ein Objekt den Trigger verlässt.
	 */
    void OnTriggerExit( Collider other )
    {
		if (!IsClose)
		{
			// Ist es ein Spieler soll die Animation gespielt werden.
			if (Config.IsStandalone && other.tag.Equals("Player")) {
				RPCAnimate(false);
			}
			else if (Config.IsServer){
				Close ();
			}
			IsClose = true;
		}
    }

    #endregion
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS  --------------------------------------------------------------------
 
    private void MakePermanent()
    {
        IsInit = true;
     
        if( Type == TriggerType.PermanentOpen ) Open();
        else if( Type == TriggerType.InstantOpen )
        {
            NetView.RPC( "RPCAnimateInstant", RPCMode.All, true );
        }
    }
 
 	/**
 	 * Diese Methode wird aufgerufen, wenn es sich um eine Cave handelt und 
 	 * der aktuelle Rechner der Server ist. Hier wird der RPC [Remote Procedure Call]
 	 * aufgerufen. Die dann die Methode RPCAnimate auf allen Rechnern mit den Parameter
 	 * true aufruft.
 	 */
    private void Open()
    {
        networkView.RPC( "RPCAnimate", RPCMode.All, true );
    }
 
 	/**
 	 * Diese Methode wird aufgerufen, wenn es sich um eine Cave handelt und 
 	 * der aktuelle Rechner der Server ist. Hier wird der RPC [Remote Procedure Call]
 	 * aufgerufen. Die dann die Methode RPCAnimate auf allen Rechnern mit den Parameter
 	 * false aufruft.
 	 */
    private void Close()
    {
        NetView.RPC( "RPCAnimate", RPCMode.All, false );
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS RPC -----------------------------------------------------------------
 
	/**
	 * Diese Methode wird auf allen Rechnern ausgeführt und wird gestartet als RPC vom
	 * Server aus.
	 * 
	 * @param open True, wenn die Tür geöffnet werden soll. Andernfalls wird sie geschlossen. 
	 */
    [RPC]
    void RPCAnimate( bool open )
    {
        if( open )
        {
            foreach( AnimationClip clip in AnimationsOpen ) {
                Ani.PlayQueued( clip.name );
			}
        }
        else
        {
            foreach( AnimationClip clip in AnimationsClose )
                Ani.PlayQueued( clip.name );
        }
    }
 
 
    [RPC]
    void RPCAnimateInstant( bool open )
    {
        if( open )
        {
            Ani.Play( AnimationsOpen[0].name );
        }
        else
        {
            Ani.Play( AnimationsClose[0].name );
        }
    }
    // =============================================================================
}

File: ISS Emergency/Assets/Scripts/TriggerHatch.js
#pragma strict

// Hannes Helmholz
//
// class for animated objects with one open and one close animation


@script RequireComponent(NetworkView) // for RPC

public class TriggerHatch extends MonoBehaviour {

	// =============================================================================
	// MEMBERS ---------------------------------------------------------------------
	public enum TriggerType {Trigger, InstantOpen, PermanentOpen, PermanentClosed}
	public enum TriggerObject {DistanceByCollider, Script}
	
	public var Type:TriggerType;
	public var Trigger:TriggerObject;
	public var AnimationsOpenSize:int = 1;
	public var AnimationsCloseSize:int = 1;
	public var AnimationsOpen:AnimationClip[] = new AnimationClip[AnimationsOpenSize];
	public var AnimationsClose:AnimationClip[] = new AnimationClip[AnimationsCloseSize];
	public var Script:TriggerButtonSequence;
	
	private var Config:ConfigClass;
	private var Ani:Animation;
	private var NetView:NetworkView;
	private var IsInit:boolean = false;
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS UNITY ---------------------------------------------------------------
	
	function Start() {
		Config = GameObject.FindWithTag("Config").GetComponent.<ConfigClass>();
		Ani = animation;
		NetView = networkView;
		
		if (!Config.IsServer) {
			IsInit = true;
			return;
		}
		
		if (Type == TriggerType.Trigger)
			IsInit = true;
			
		if (Trigger == TriggerObject.Script)
			Script = transform.GetComponent.<TriggerButtonSequence>();
	}
	
	
	function Update() {	
		if (!Config.IsServer)
			return;
			
		if (!IsInit && Config.IsGameStarted)
			MakePermanent();
	}
	
	
	function OnTriggerEnter(other:Collider) {
		if (Trigger != TriggerObject.DistanceByCollider || !Config.IsServer || !other.gameObject.tag.Equals("Player"))
			return;	
				
		Open();
	}
	
	
	function OnTriggerExit(other:Collider) {
		if (Trigger != TriggerObject.DistanceByCollider || !Config.IsServer || !other.gameObject.tag.Equals("Player"))
			return;
			
		Close();
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS  --------------------------------------------------------------------
	
	private function MakePermanent() {
		IsInit = true;
		
		if (Type == TriggerType.PermanentOpen)
			Open();
		else if (Type == TriggerType.InstantOpen)
			NetView.RPC("RPCAnimateInstant", RPCMode.AllBuffered, true);
	}
	
	
	private function Open() {
		NetView.RPC("RPCAnimate", RPCMode.AllBuffered, true);
	}
	
	
	private function Close() {
		NetView.RPC("RPCAnimate", RPCMode.AllBuffered, false);
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS RPC -----------------------------------------------------------------
	
	@RPC
	function RPCAnimate(open:boolean) {
		if (open)
			for (var clip:AnimationClip in AnimationsOpen)
				Ani.PlayQueued(clip.name);
		else
			for (var clip:AnimationClip in AnimationsClose)
				Ani.PlayQueued(clip.name);
	}
	
	
	@RPC
	function RPCAnimateInstant(open:boolean) {
		if (open)
			Ani.Play(AnimationsOpen[AnimationsOpen.Length - 1].name);
		else
			Ani.Play(AnimationsClose[AnimationsClose.Length - 1].name);
	}
	// =============================================================================
}